### ISSUE TITLE
Delegate= cannot work with unified hierachy as designed

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#7355]
I started writing a commit message, and then realized that this cannot work without changes:

With v235-249-g0263828039, when starting services with Delegate=yes on unified cgroup hierarchy, we'd fail the "no internal process constraint". Specifically, we'd do the following (strace from pid1):
    
    mkdir("/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service", 0755) = 0
    inotify_add_watch(6, "/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.events", IN_MODIFY) = 1297
    openat(AT_FDCWD, "/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control", O_WRONLY|O_NOCTTY|O_CLOEXEC) = 94
    write(94, "+io\n", 4)                   = 4
    close(94)                               = 0
    openat(AT_FDCWD, "/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control", O_WRONLY|O_NOCTTY|O_CLOEXEC) = 94
    write(94, "+memory\n", 8)               = 8
    close(94)                               = 0
    openat(AT_FDCWD, "/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cgroup.subtree_control", O_WRONLY|O_NOCTTY|O_CLOEXEC) = 94
    write(94, "+pids\n", 6)                 = 6
    close(94)                               = 0
    ...
    clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7fd5a79facd0) = 11682
    ...
    write(94, "11682\n", 6)     = -1 EBUSY (Device or resource busy)
    
PIDs cannot be written to a cgroup which also has the `cgroup.subtree_control` file populated. We need to leave the job of enabling the controllers to the delegate. It needs to first move itself to a subgroup (e.g. `<delegation-root>/init.scope`), and then turn on controllers in `<delegation-root>/cgroup.subtree_control`.
    
When running on hybrid hierarchy, the problem wasn't seen because the v2 hierarchy we use for managing processes doesn't have any controllers enabled.

To actually limit the controllers delegated to the child, we need effect `cgroup.controllers` with a subset of the controllers. One way to do this is to insert an extra cgroup above the delegate cgroup, so we'd have e.g. `/sys/fs/cgroup/user.slice/user-1000.slice/wrapper/user@1000.service`, and populate `/sys/fs/cgroup/user.slice/user-1000.slice/wrapper/cgroup.subtree_control` with the list of delegated controllers.

/cc @poettering, @htejun

**Comment by keszybz** [comment]
Also, Delegate=something (not "no" and not "yes") would not work in the sense that the delegate has write permissions to the `cgroup.subtree_control` file so they can enable/disable controllers at will.

**Comment by keszybz** [comment]
FTR, this makes the user@.service work again, but it breaks Delegate=something:
```diff
diff --git src/core/cgroup.c src/core/cgroup.c
index 3c0d22a691..7cf864bdf6 100644
--- src/core/cgroup.c
+++ src/core/cgroup.c
@@ -1072,36 +1072,26 @@ CGroupMask unit_get_own_mask(Unit *u) {
         if (!c)
                 return 0;
 
-        return cgroup_context_get_mask(c);
-}
-
-CGroupMask unit_get_delegate_mask(Unit *u) {
-        CGroupContext *c;
-
         /* If delegation is turned on, then turn on selected controllers, unless we are on the legacy hierarchy and the
          * process we fork into is known to drop privileges, and hence shouldn't get access to the controllers.
          *
          * Note that on the unified hierarchy it is safe to delegate controllers to unprivileged services. */
 
-        if (u->type == UNIT_SLICE)
-                return 0;
-
-        c = unit_get_cgroup_context(u);
-        if (!c)
-                return 0;
-
-        if (!c->delegate)
-                return 0;
-
-        if (cg_all_unified() <= 0) {
-                ExecContext *e;
-
-                e = unit_get_exec_context(u);
-                if (e && !exec_context_maintains_privileges(e))
-                        return 0;
+        if (c->delegate) {
+                if (cg_all_unified() > 0)
+                        return c->delegate_controllers;
+                else {
+                        ExecContext *e;
+
+                        e = unit_get_exec_context(u);
+                        assert(e); /* delegate can only be set for service/socket/mount/swap units */
+
+                        if (exec_context_maintains_privileges(e))
+                                return c->delegate_controllers;
+                }
         }
 
-        return c->delegate_controllers;
+        return cgroup_context_get_mask(c);
 }
 
 CGroupMask unit_get_members_mask(Unit *u) {
@@ -1112,7 +1102,7 @@ CGroupMask unit_get_members_mask(Unit *u) {
         if (u->cgroup_members_mask_valid)
                 return u->cgroup_members_mask;
 
-        u->cgroup_members_mask = unit_get_delegate_mask(u);
+        u->cgroup_members_mask = 0;
 
         if (u->type == UNIT_SLICE) {
                 void *v;
@@ -1127,7 +1117,8 @@ CGroupMask unit_get_members_mask(Unit *u) {
                         if (UNIT_DEREF(member->slice) != u)
                                 continue;
 
-                        u->cgroup_members_mask |= unit_get_subtree_mask(member); /* note that this calls ourselves again, for the children */
+                        /* note that this calls ourselves again, for the children */
+                        u->cgroup_members_mask |= unit_get_subtree_mask(member);
                 }
         }
 
diff --git src/core/cgroup.h src/core/cgroup.h
index a75be38044..1b5807c518 100644
--- src/core/cgroup.h
+++ src/core/cgroup.h
@@ -154,7 +154,6 @@ void cgroup_context_free_blockio_device_weight(CGroupContext *c, CGroupBlockIODe
 void cgroup_context_free_blockio_device_bandwidth(CGroupContext *c, CGroupBlockIODeviceBandwidth *b);
 
 CGroupMask unit_get_own_mask(Unit *u);
-CGroupMask unit_get_delegate_mask(Unit *u);
 CGroupMask unit_get_members_mask(Unit *u);
 CGroupMask unit_get_siblings_mask(Unit *u);
 CGroupMask unit_get_subtree_mask(Unit *u);
```

**Comment by poettering** [comment]
Fix waiting in #7381
